packages/playwright-core/src/server/injected/ariaSnapshot.ts

```diff
@@ -248,49 +248,65 @@ export function matchesAriaTree(builtins: Builtins, rootElement: Element, templa
 
 export function getAllByAria(builtins: Builtins, rootElement: Element, template: AriaTemplateNode): Element[] {
   const root = generateAriaTree(builtins, rootElement, 0, false).root;
-  const matches = matchesNodeDeep(root, template, true);
+  const matches = matchesNodeDeep(root, template, true, false);
   return matches.map(n => n.element);
 }
 
-function matchesNode(node: AriaNode | string, template: AriaTemplateNode, depth: number): boolean {
+function matchesNode(node: AriaNode | string, template: AriaTemplateNode, isDeepEqual: boolean): boolean {
   if (typeof node === 'string' && template.kind === 'text')
     return matchesTextNode(node, template);
 
-  if (node !== null && typeof node === 'object' && template.kind === 'role') {
-    if (template.role !== 'fragment' && template.role !== node.role)
-      return false;
-    if (template.checked !== undefined && template.checked !== node.checked)
-      return false;
-    if (template.disabled !== undefined && template.disabled !== node.disabled)
-      return false;
-    if (template.expanded !== undefined && template.expanded !== node.expanded)
-      return false;
-    if (template.level !== undefined && template.level !== node.level)
-      return false;
-    if (template.pressed !== undefined && template.pressed !== node.pressed)
-      return false;
-    if (template.selected !== undefined && template.selected !== node.selected)
-      return false;
-    if (!matchesName(node.name, template))
-      return false;
-    if (!matchesText(node.props.url, template.props?.url))
-      return false;
-    if (!containsList(node.children || [], template.children || [], depth))
+  if (node === null || typeof node !== 'object' || template.kind !== 'role')
+    return false;
+
+  if (template.role !== 'fragment' && template.role !== node.role)
+    return false;
+  if (template.checked !== undefined && template.checked !== node.checked)
+    return false;
+  if (template.disabled !== undefined && template.disabled !== node.disabled)
+    return false;
+  if (template.expanded !== undefined && template.expanded !== node.expanded)
+    return false;
+  if (template.level !== undefined && template.level !== node.level)
+    return false;
+  if (template.pressed !== undefined && template.pressed !== node.pressed)
+    return false;
+  if (template.selected !== undefined && template.selected !== node.selected)
+    return false;
+  if (!matchesName(node.name, template))
+    return false;
+  if (!matchesText(node.props.url, template.props?.url))
+    return false;
+
+  // Proceed based on the container mode.
+  if (template.containerMode === 'contain')
+    return containsList(node.children || [], template.children || []);
```

Comment:
If we had a parent with deepEqual, should we inherit that to our grandchildren, i.e. only opt-out this particular node? As implemented, setting `contain` resets `deep-equal` for the whole subtree. Perhaps `deep-contain` is justified?

Reply:
I'd say they need to opt into deep equal again for that. That way you only need to climb to the nearest directive to understand the policy.